libkatana_python_native/src/PropertyGraph.cpp

```diff
@@ -0,0 +1,742 @@
+    }
+  }
+
+  if (!arrow_table) {
+    return KATANA_ERROR(
+        katana::ErrorCode::InvalidArgument,
+        "A table argument (a dict or arrow table) or a keyword argument must "
+        "be provided.");
+  }
+
+  return arrow_table;
+}
+
+class LazyDataAccessor {
+public:
+  virtual ~LazyDataAccessor();
+  virtual py::object at(ssize_t i) const = 0;
+  virtual py::object array(const py::slice& slice) const = 0;
+};
+
+LazyDataAccessor::~LazyDataAccessor() = default;
+
+template <typename T>
+class LazyDataAccessorTyped : public LazyDataAccessor {
+public:
+  virtual T at_typed(ssize_t i) const = 0;
+  virtual py::object at(ssize_t i) const final { return py::cast(at_typed(i)); }
+  virtual py::object array(const py::slice& slice) const {
+    auto begin = py::cast<ssize_t>(slice.attr("start"));
+    auto end = py::cast<ssize_t>(slice.attr("stop"));
+    auto step = py::cast<ssize_t>(slice.attr("step"));
+    py::array_t<T> out{(end - begin) / step};
+    for (ssize_t j = 0, i = begin; i < end; j += 1, i += step) {
+      out.mutable_at(j) = at_typed(i);
+    }
+    return std::move(out);
+  }
+};
+
+class GraphBaseEdgeDestAccessor final
+    : public LazyDataAccessorTyped<katana::PropertyGraph::Node> {
+  const katana::PropertyGraph* pg_;
+
+public:
+  virtual ~GraphBaseEdgeDestAccessor();
+
+  GraphBaseEdgeDestAccessor(const katana::PropertyGraph* pg) : pg_(pg) {}
+
+  virtual katana::PropertyGraph::Node at_typed(ssize_t i) const {
+    return *pg_->OutEdgeDst(i);
+  }
+};
+GraphBaseEdgeDestAccessor::~GraphBaseEdgeDestAccessor() = default;
+
+class GraphBaseEdgeSourceAccessor final
+    : public LazyDataAccessorTyped<katana::PropertyGraph::Node> {
+  const katana::PropertyGraphViews::BiDirectional view_;
+
+public:
+  virtual ~GraphBaseEdgeSourceAccessor();
+
+  GraphBaseEdgeSourceAccessor(katana::PropertyGraph* pg)
+      : view_(pg->BuildView<katana::PropertyGraphViews::BiDirectional>()) {}
+
+  virtual katana::PropertyGraph::Node at_typed(ssize_t i) const {
+    return view_.GetEdgeSrc(i);
+  }
+};
+GraphBaseEdgeSourceAccessor::~GraphBaseEdgeSourceAccessor() = default;
+
+template <typename T>
+py::object
+range(T start, T end, T step = 1) {
+  py::module builtins = py::module::import("builtins");
+  return builtins.attr("range")(start, end, step);
+}
+
+namespace {
+katana::GraphTopologyTypes::Edge
+out_edge_index(katana::PropertyGraph* pg, katana::GraphTopologyTypes::Node n) {
+  return *pg->OutEdges(n).begin();
+}
+katana::GraphTopologyTypes::Node
+out_edge_dst(katana::PropertyGraph* pg, katana::GraphTopologyTypes::Edge e) {
+  return *pg->OutEdgeDst(e);
+}
+}  // namespace
+
+void
+DefPropertyGraph(py::module& m) {
+  using namespace katana;
+  using namespace katana::python;
+
+  py::class_<PropertyGraph, std::shared_ptr<PropertyGraph>> cls(m, "Graph");
+
+  katana::DefCythonSupport(cls);
+  katana::DefConventions(cls);
+  katana::RegisterNumbaClass(cls);
+
+  cls.def(
+      py::init(
+          [](py::object path,
```

Comment:
Why does this have to be a `py::object`? Can't this just be a `py::str` directly? Or is this so we can catch type errors when casting? Or if we pass in paths like `url.URL` or `pathlib.Path` that have string representations?

Reply:
Yes. I want this to be able to accept `Path`.